packages/babel-generator/src/generators/statements.ts

```diff
@@ -82,16 +82,26 @@ export function WhileStatement(this: Printer, node: t.WhileStatement) {
   this.printBlock(node);
 }
 
-const buildForXStatement = function (op) {
-  return function (node: any) {
+const buildForXStatement = function (op: "in" | "of") {
+  return function (this: Printer, node: t.ForInStatement | t.ForOfStatement) {
     this.word("for");
     this.space();
-    if (op === "of" && node.await) {
+    if (op === "of" && (node as t.ForOfStatement).await) {
       this.word("await");
       this.space();
     }
     this.token("(");
+
+    // ECMAScript specifically forbids a for-of loop from starting with the
+    // token sequence "for ( async of", because it would be ambiguous with
+    // "for (async of => {};;)", so we need to add extra parentheses.
+    const leftNeedsParens =
+      op === "of" && t.isIdentifier(node.left) && node.left.name === "async";
+
+    if (leftNeedsParens) this.token("(");
```

Comment:
We have `packages/babel-generator/src/node/parentheses.ts` for determining whether an AST node should be wrapped in parentheses.

You can add a new `Identifier` handler and check if the `parent` is a forOfStatement and the node comes from left.